### ISSUE TITLE
Add a sensor chart box for the dashboard

### ISSUE THREAD

**Issue by Pierre-Gilles** [GladysAssistant/Gladys#693]
**Description**

Be able to display a chart on the dashboard. Ex: temperature of last month

**Specs**

- [ ] User can create a chart box on the dashboard
- [ ] The user select the device feature he wants to display in the list of "read_only" device feature. Only interesting device features should be displayed. **Binary should not be displayed for ex**
- [ ]  The user select a range between a select box. (Last 24h, last 2 days, last week, last month)
- [ ] The box displays the chart, he a nice way. Take inspiration on the homepage of our theme => https://preview.tabler.io/

**Comment by euguuu** [comment]
it's ok for you if I try to do this ticket or you already start dev for this ?

**Comment by VonOx** [comment]
@atrovato is assigned from 16 May but no PR , I think you can !

**Comment by atrovato** [comment]
I confirm, I let you this one.

**Comment by Pierre-Gilles** [comment]
Let's talk about the chart library you use before starting the dev ! 

I'm really in favor of using small (not too many deps) and well maintained library for the frontend.

I like https://bundlephobia.com/ to measure a library size.

**Comment by atrovato** [comment]
I don't have anything to share on this, but we already use a chart lib, maybe this is enough https://visjs.org/

**Comment by Pierre-Gilles** [comment]
Honestly I don't really like this lib :D I want to remove it from Gladys ^^ 

In Preact we do code splitting per page, so the visjs file is only downloaded when you enter the "Z-Wave / network" page. I don't think we want this big library on the dashboard which is the main page.

I think we should cherry-pick a tiny library for this to avoid huge page load on the first page.

**Comment by Pierre-Gilles** [comment]
I like chartjs: https://www.chartjs.org/docs/latest/

And it seems it's tiny (no external dependency): https://bundlephobia.com/result?p=chartjs@0.3.2

Edit: see below I mistyped chart.js

**Comment by Pierre-Gilles** [comment]
Bundle comparaison of the two:

<img width="1417" alt="Screenshot 2020-11-19 at 11 22 54" src="https://user-images.githubusercontent.com/7365207/99646783-98dcdf80-2a59-11eb-96bf-253933d29f45.png">
<img width="1419" alt="Screenshot 2020-11-19 at 11 22 34" src="https://user-images.githubusercontent.com/7365207/99646793-9bd7d000-2a59-11eb-8878-137178ba2402.png">

**Comment by Pierre-Gilles** [comment]
Ok my bad, I typed the wrong module on bundlephobia, I type "chartjs" instead of "chart.js"... 

chart.js is big:

<img width="1422" alt="Screenshot 2020-11-19 at 11 27 17" src="https://user-images.githubusercontent.com/7365207/99647258-3506e680-2a5a-11eb-8ac4-f68a8503e9c5.png">

**Comment by Pierre-Gilles** [comment]
There is the famous d3.js library which is a little big too but I read that the library is tree-shakable, which means that maybe it could be possible to import only what we need in the lib! ( [Read more about Tree-shaking](https://webpack.js.org/guides/tree-shaking/) ) 

I found victory too: https://github.com/FormidableLabs/victory

And it's tree-shakable ! 

<img width="1406" alt="Screenshot 2020-11-19 at 11 38 11" src="https://user-images.githubusercontent.com/7365207/99648628-ece8c380-2a5b-11eb-8667-0a270660746f.png">

**Comment by Pierre-Gilles** [comment]
I really like nivo too: https://nivo.rocks/line/

We could even use it do to presence/absence, heating on/off tracking in Gladys: 

<img width="723" alt="Screenshot 2020-11-19 at 11 42 42" src="https://user-images.githubusercontent.com/7365207/99649009-5bc61c80-2a5c-11eb-8e78-51b1a749d793.png">

**Comment by Pierre-Gilles** [comment]
Ok I remembered that the theme we are using in Gladys 4 (tabler) is doing chart as well!

Example:

<img width="1440" alt="Screenshot 2020-11-19 at 11 45 04" src="https://user-images.githubusercontent.com/7365207/99649482-ee66bb80-2a5c-11eb-9222-9280f3539aa5.png">

Example 2:

<img width="681" alt="Screenshot 2020-11-19 at 11 45 07" src="https://user-images.githubusercontent.com/7365207/99649563-050d1280-2a5d-11eb-880d-a4838f035eda.png">

I think they are using https://apexcharts.com/ https://github.com/apexcharts/apexcharts.js, which is open-source too.

It looks nice!

**Comment by Pierre-Gilles** [comment]
To recap: 

| Library | Pros | Cons |
| ------------- | ------------- | ------------- |  
| ApexCharts  | Well integrated with our theme. Really beautiful. Lots of charts possible. | 118.2kB bundle size gzipped
| Chart.js  | Beautiful  & rock solid. Been here for a long time |  123.8 kB size gzipped
| Nivo  | Beautiful & large collection of charts. Each component can be loaded independently. 83.8 kB gzipped for the line component. Tree-shakable |  Huge library to understand. I didn't get if 83kb is enough to have the whole line library working?  
| Victory  | Tree-shakable. Large range of charts available | Not beautiful out of the box, we need to customize it

This is my first results after investigating during 45 minutes, I'm open for recommandation which fits our needs :) 

Also, testing each lib can be a good idea to see the impact/ease of use of each of them.

**Comment by euguuu** [comment]
In my opinion, charts integration in gladys matches to:

principal :
* display historical data of home sensors 
* display historical data of user health 

secondary:
* display interaction between domotic hardware (like network in zwave)
* help to optimize home management (like electricity consumption)
* help to optimize health management (like sport session, km in time for runner)

for the moment feature state coresponds to timeseries data , I am not sure this point needs to evoluate,
maybe position used with owntrack and user presence can open another data point of view 

for those needs I think those chart types match:
* line / area
* column / bar
* timeline
* pie
* network
* bubble

=> I don't see network chart type allowed in chartjs, apexcharts and victory 
=> Nivo has all we need

In terms of documentation, 
Nivo appears more complicated (no big documentation)
chartjs, apexcharts and victory seems good (big documentation and get started seems easy)

In terms of look & feel,
apexcharts default view appears closest to gladys 
victory and nivo are very distant to gladys theme => I think to much effort need to build something like ladys (except if a real designer is available in gladys community)

So for me the winner is apexcharts (but chartjs is just behind and perenity of this lib can be an important point)
for network chart we cn continue with visjs if it's impossible with apexcharts or chartjs.

**Comment by Pierre-Gilles** [comment]
> So for me the winner is apexcharts (but chartjs is just behind and perenity of this lib can be an important point)

I agree with you. Can you do a quick proof-of-concept of apexcharts? (in a separate PR)

By the way, just a point on the feature. You'll have to do a backend route to get device feature states. Be careful not to return all values, you can do something called "downsampling" ([Wikipedia](https://en.wikipedia.org/wiki/Downsampling_(signal_processing)) if you are not familiar with the concept)

For example, if the user wants to display the last week of data and his sensors is saving 1 value/minute, it means there is 10k values in the database for 1 week of data, and it's probably too much to display on the frontend + it'll take too much time to download a 10k rows array in the API.

So you can use something like this NPM package (https://www.npmjs.com/package/downsample, to be tested) backend-side  to reduce the 10k elements array to something with less value (like 50? 100? to be tested in the UI). 

Does it make sense to you ? :) 

If you have question don't hesitate. The NPM package I recommended for downsampling is just the first one I found on NPM, so if you find a better one (more stars, more usage), you can definitely use another one :)

**Comment by euguuu** [comment]
ok

**Comment by euguuu** [comment]
how can I change the status of this issue to put it 'in progress' ?

**Comment by Pierre-Gilles** [comment]
Maybe simply create a WIP PR with a link to this issue ? It'll display a message in this issue

**Comment by VonOx** [comment]
You are assignee and for linking issue to PR you use theses keywords

https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Comment by euguuu** [comment]
I put a PR linked to this issue with first version of poc (you need t put a 'device in room chart' box in edit mode, no need to specify something in form all data are fake)

![image](https://user-images.githubusercontent.com/9742965/99687242-b5dad800-2a84-11eb-9cea-9b6dafd92c3f.png)

![image](https://user-images.githubusercontent.com/9742965/99687286-c4c18a80-2a84-11eb-88d8-ab773fef6dcf.png)

**Comment by Pierre-Gilles** [comment]
Nice! 

Can you see if you can get the same beautiful design we can see on tabler.io website? (like here: https://github.com/GladysAssistant/Gladys/issues/693#issuecomment-730256467)

**Comment by euguuu** [comment]
in fact is very easy to do same design of tabler.io because it uses apexchart to display graph so it's just a copy of chart options to do the same

example integrate in gladys:
![image](https://user-images.githubusercontent.com/9742965/99790333-3dc3ef00-2b24-11eb-94ca-38c7417478fa.png)

![image](https://user-images.githubusercontent.com/9742965/99790366-49171a80-2b24-11eb-9832-1e68399277c1.png)

![image](https://user-images.githubusercontent.com/9742965/99790403-56cca000-2b24-11eb-897b-afdec3c425d6.png)

**Comment by Pierre-Gilles** [comment]
I was talking about this design: 

<img width="681" alt="99649563-050d1280-2a5d-11eb-880d-a4838f035eda" src="https://user-images.githubusercontent.com/7365207/99800723-7623f700-2b3d-11eb-9c44-675b48973aaa.png">

Can't we do something more "edge to edge" ?

**Comment by euguuu** [comment]
this design is the same of area screenshot I do (without line in background and no toolbar), the box with "Today's Earning ..." is just HTML component, not a part of chart, so it's possible to do that I work on it for next step of this dev.

**Comment by euguuu** [comment]
I try to do this render but I have some issue

I try to create 2 types of box : 
1- one with only 1 feaure display
2- one with multiple features display

1- I try to reproduce 
![image](https://user-images.githubusercontent.com/9742965/100103917-ce246b80-2e65-11eb-8f8b-beea47b1bbb0.png)

but for the moment I do this:
![image](https://user-images.githubusercontent.com/9742965/100103851-b947d800-2e65-11eb-958a-22d00948a8a6.png)

=> I don't understand why but I can't change height of chart and dropdown list doesn't work

2- This box is just to display any features for any room so I want do the same box with legend and a "fullscreen" link, in first step I try to factorize code and apply same method of deviceInRoomBox but I don't understand why chart don't refresh data (Ithink I miss something in life cycle of props/state/store/blablabla...)

Somedbody can help me about those issues ?

(I need to add some config for tooltip design and add loading on box but it's not difficult but I need to make a break on front part ;-) )

**Comment by Pierre-Gilles** [comment]
Looks really nice! 

Just one feedback:

![100103851-b947d800-2e65-11eb-958a-22d00948a8a6](https://user-images.githubusercontent.com/7365207/100428088-204ed200-309c-11eb-9d87-3c82d0f10d3d.png)

> I don't understand why but I can't change height of chart and dropdown list doesn't work

So you use the React-apexchart library and the height={} props doesn't work?

> I need to make a break on front part ;-) 

Sometimes all we need is a break to see the issues when coming back :D

**Comment by euguuu** [comment]
I do some change on this issue

* For One feature chart:
** EditBox:
![image](https://user-images.githubusercontent.com/9742965/100647253-873fe580-333f-11eb-91fa-88215886f2a0.png)


** Dash Box:
![image](https://user-images.githubusercontent.com/9742965/100647189-6d060780-333f-11eb-9d77-e5a3c00f9cd7.png)


* For multi features chart:
** EditBox:
![image](https://user-images.githubusercontent.com/9742965/100648167-da666800-3340-11eb-8fa6-a9dc116afac3.png)


** Dash Box:
![image](https://user-images.githubusercontent.com/9742965/100648201-ea7e4780-3340-11eb-9943-758dfd7bfe17.png)

=> I wait your feedback to change anythings. 

actually I exclude those type of feature to feature list (for both): 
const excludeFeatyreType = [
        DEVICE_FEATURE_TYPES.SENSOR.BINARY,
        DEVICE_FEATURE_TYPES.CAMERA.IMAGE,
        DEVICE_FEATURE_TYPES.SENSOR.UNKNOWN,
        DEVICE_FEATURE_TYPES.UNKNOWN.UNKNOWN,
        DEVICE_FEATURE_TYPES.LIGHT.COLOR
      ];
=> Any others type ?

TODO:
* [optionnal] put screenshot of chart type in edit box select
* improve design of column chart ( principaly for multi feature display )
* add  link to open multi feature chart in large screen (I try to use modal button describe in this page https://preview.tabler.io/modals.html but doesn't work, there are any example of modal box in gladys ?)
* improve server test
* challenge downsample & trend algo

**Comment by VonOx** [comment]
Wow it's beautiful. 
I think binary sensor must be added, a user must see when door is open or closed over a day ( just use case). No sampling in that case

**Comment by Pierre-Gilles** [comment]
@VonOx let's keep binary sensor for later? One step at a time, we don't want a monster PR :) 

I think we want to treat binary a different way, we want to display a specific chart for that, like: 

<img width="727" alt="Screenshot 2020-12-03 at 09 40 28" src="https://user-images.githubusercontent.com/7365207/100978854-a3b06d80-354b-11eb-8941-603231c7ef83.png">

**Comment by VonOx** [comment]
Yes I agree this kind of graph is very useful, we can implement later on a good base.

**Comment by Pierre-Gilles** [comment]
@euguuu Thanks for your work it's really nice! :) 

First on the design, I like it but I think we could be more "edge to edge". Could you remove all padding on the boxes so there is no blank border?

![100647189-6d060780-333f-11eb-9d77-e5a3c00f9cd7](https://user-images.githubusercontent.com/7365207/100979136-13265d00-354c-11eb-9099-6c4f8ff800ba.png)


For the exclude list, you can exclude the cube too: 

```
CUBE: {
    MODE: 'mode',
    ROTATION: 'rotation',
  },
```

And the button too: 

```
BUTTON: {
    CLICK: 'click',
  },
```

What's the "choose subset of devices" ? I don't find the UI very explicit. 

> there are any example of modal box in gladys ?)

We don't do modals in Gladys as it's not really a good practice when your Webapp is also a mobile app. Modals looks great on desktop (it looks like windows in a windows), but on mobile it makes the flow complicated. For now, I haven't seen any use case that was really requiring a modal. 

> challenge downsample & trend algo

Which algo did you use at the end?

**Comment by euguuu** [comment]
for reduce number of object returns I use downsample: https://www.npmjs.com/package/downsample  , the cgladys ontroller can accept a parameter to choose algo (ASAP , SMA , LTTB , LTOB , LTD ), now I use LTTB with 1000 objects , I need more time to test all algo and choose the best for representation/performance and maybe adapt default number values return 

for trend (only the little green or red arrow => one value return to front)  I use https://www.npmjs.com/package/trend , 
I do'nt find better one, I'm not sure is very usefull in this box

**Comment by Pierre-Gilles** [comment]
> for reduce number of object returns I use downsample: https://www.npmjs.com/package/downsample , the controller can accept a parameter to choose algo (ASAP , SMA , LTTB , LTOB , LTD ), now I use LTTB with 1000 objects 

Ok! When you find the best algorithm, just set it in your function and remove the controller parameter, this is not really something needed client side, let's keep the API as clear as possible and with the best default :) 

> for trend (only the little green or red arrow => one value return to front) I use https://www.npmjs.com/package/trend 

I like the trend and I think we want to do keep it, but this library is not something we can keep in Gladys, it's outdated and not supported anymore: 

<img width="429" alt="Screenshot 2020-12-03 at 10 09 16" src="https://user-images.githubusercontent.com/7365207/100981712-d5c3ce80-354f-11eb-99e2-45f6c79a82af.png">

I think the trend is something we should calculate ourself.

Let's get back to the use case: When we display a trend, the goal is to understand how the value is evolving.

**Example:** 

The humidity is 90%  right now, it was 70% one hour ago. 

V% = (90-70)/70*100 = +29%

Now the main question is: which values are we comparing? 
 
I think we could compare the first value of the interval displayed by the user with the last value of the interval (the current value of the sensor)

If the user displays the past hour of temperature on his dashboard, it means we could compare the value from 1 hour ago (first value of the chart), to the current value (last value of the chart)

What do you think of that?

**Comment by euguuu** [comment]
### **Downsample tests:**
All of the algorithm use a chartWidth to limit number of results but it's not a strict limit for ASAP (result is very close to limit) & SMA (result is very far to limit)

**Algo theory:**
- LTOB:This algorithm is very simple. First all the points are ranked by calculating theireffective areas. Points with effective areas as null are excluded. The data pointsare then split up into approximately equal number of buckets as the specified down-sample threshold. Finally, one point with the highest rank (largest effective area) isselected to represent each bucket in the downsampled data.
- LTTB: same of LTOB but treat 3 points in comparison area (improve precision)
- LTD: same of lTTB but the range for comparison area is calculated dynamicaly (improve precision but more time need on calcutation [server side))
- SMA: In statistics, a moving average (rolling average or running average) is a calculation to analyze data points by creating a series of averages of different subsets of the full data set. It is also called a moving mean (MM) or rolling mean and is a type of finite impulse response filter. Variations include: simple, and cumulative, or weighted forms (described below). 
=> I'm not really understand the variation style use in this algo (I try to do more test with different chartWidth to understand)
=> normally this algo must be goo for representation but very difficult to ensure the number of results returns
- ASAP: automatically smoothes time series plots to remove short-term noise while retaining large-scale deviations.

### TESTS:
I try 2 different dataset with 1 value by minutes => 44 640 values for 31 days 
(first size value is transfer value and second value after decompression I suppose, extract with console of firefox)
(with 44 640 values I never see representation , I don't undestand why I try to fix this)

**Test 1:**
Dataset : This dataset use random value between 1 and 50 (worst case for representation)

- LTOB:
chartWidth 1000 : 6,21 Ko (taille 39,75 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288230-bc43b080-37f5-11eb-9e5c-5e3325b57b32.png)

chartWidth 500 : 3,75 Ko (taille 20,41 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288236-c1a0fb00-37f5-11eb-87b4-e5eb3fe67159.png)

chartWidth 250 : 2,48 Ko (taille 10,79 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288237-c49beb80-37f5-11eb-828e-a853a14ebe52.png)

chartWidth 100 : 1,63 Ko (taille 5,00 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288240-c82f7280-37f5-11eb-878b-5a9ef2343b19.png)

chartWidth 50 : 1,31 Ko (taille 3,07 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288242-ca91cc80-37f5-11eb-80a3-49586d85d336.png)

- LTTB:
chartWidth 1000 : 6,09 Ko (taille 39,72 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288295-1fcdde00-37f6-11eb-8240-0ef66a3041e6.png)

chartWidth 500 : 3,68 Ko (taille 20,43 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288294-1cd2ed80-37f6-11eb-84af-9b6137494d2b.png)

chartWidth 250 : 2,28 Ko (taille 10,77 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288291-1a709380-37f6-11eb-8ded-2a10f5b159dd.png)

chartWidth 100 : 1,58 Ko (taille 5,00 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288285-16dd0c80-37f6-11eb-8a01-6e165bd5203d.png)

chartWidth 50 : 1,30 Ko (taille 3,07 Ko)

- LTD:
chartWidth 1000 : 6,12 Ko (taille 39,72 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288350-66233d00-37f6-11eb-89bb-f428b665801b.png)

chartWidth 500 : 3,66 Ko (taille 20,43 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288354-6b808780-37f6-11eb-8de8-50b92e30cefa.png)

chartWidth 250 : 2,29 Ko (taille 10,79 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288357-6faca500-37f6-11eb-9be9-86c6851e260e.png)

chartWidth 100 : 1,58 Ko (taille 5 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288363-73402c00-37f6-11eb-8437-ce55d4f71ef5.png)

chartWidth 50 : 1,30 Ko (taille 3,07 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288366-75a28600-37f6-11eb-978f-d5e13b3d1b45.png)

- SMA:
chartWidth 1000 : 190,68 Ko (taille 1,29 Mo) => 43641 values returns
![image](https://user-images.githubusercontent.com/9742965/101288407-a71b5180-37f6-11eb-8972-5b50684ff6b7.png)

chartWidth 500 : 192,91 Ko (taille 1,30 Mo) => 44141 values returns
![image](https://user-images.githubusercontent.com/9742965/101288400-a4b8f780-37f6-11eb-9e0d-b6d3a72a39a6.png)

chartWidth 250 : 195,04 Ko (taille 1,30 Mo) => 44391 values returns
![image](https://user-images.githubusercontent.com/9742965/101288397-a1be0700-37f6-11eb-990f-365d8a7fc21e.png)

chartWidth 100 : 193,17 Ko (taille 1,27 Mo) => 44591 values returns
![image](https://user-images.githubusercontent.com/9742965/101288394-9ec31680-37f6-11eb-9d8a-ed91664c0ac9.png)

chartWidth 50 : 189,01 Ko (taille 1,27 Mo) => 44591 values returns
![image](https://user-images.githubusercontent.com/9742965/101288389-9bc82600-37f6-11eb-9259-d538f48360fb.png)

- ASAP:
chartWidth 1000 : 8,25 Ko (taille 39,22 Ko) => 915 values returns
![image](https://user-images.githubusercontent.com/9742965/101288437-ce721e80-37f6-11eb-95e3-6b9847847516.png)

chartWidth 500 : 4,85 Ko (taille 19,92 Ko) => 452 values returns
![image](https://user-images.githubusercontent.com/9742965/101288432-cb772e00-37f6-11eb-86cc-332dbf861095.png)

chartWidth 250 : 3,04 Ko (taille 10,54 Ko) => 226 values returns
![image](https://user-images.githubusercontent.com/9742965/101288429-c914d400-37f6-11eb-84dd-ccd01bcb16da.png)

chartWidth 100 : 1,96 Ko (taille 4,92 Ko) => 91 values returns
![image](https://user-images.githubusercontent.com/9742965/101288426-c74b1080-37f6-11eb-995a-bcbab6f2cfdc.png)

chartWidth 50 : 1,50 Ko (taille 3,06 Ko) => 46 values returns
![image](https://user-images.githubusercontent.com/9742965/101288424-c4e8b680-37f6-11eb-8c8c-0ec7a283946e.png)

**Test 2:**
Dataset : This dataset use increase value between 1 and 50 (value return to 1 evrey 50 min)

- LTOB:
chartWidth 1000 : 4,14 Ko (taille 40,12 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288481-290b7a80-37f7-11eb-8936-1b07449a1661.png)

chartWidth 500 : 2,67 Ko (taille 20,67 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288477-26a92080-37f7-11eb-8802-7f1432bcbe54.png)

chartWidth 250 : 2,00 Ko (taille 10,91 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288475-2446c680-37f7-11eb-9986-282e23ba56c6.png)

chartWidth 100 : 1,43 Ko (taille 5,05 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288473-21e46c80-37f7-11eb-9174-fda16e16213f.png)

chartWidth 50 : 1,20 Ko (taille 3,10 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288467-18f39b00-37f7-11eb-9e80-0e7bac320c6e.png)

- LTTB:
chartWidth 1000 : 4,64 Ko (taille 39,74 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288501-4fc9b100-37f7-11eb-8fbf-8504abd73ea7.png)

chartWidth 500 : 2,94 Ko (taille 20,43 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288499-4ccec080-37f7-11eb-87b2-e6015853e0e2.png)

chartWidth 250 : 1,85 Ko (taille 10,79 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288497-4a6c6680-37f7-11eb-9519-25b7a25630c8.png)

chartWidth 100 : 1,43 Ko (taille 5,00 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288492-480a0c80-37f7-11eb-8d73-0f429c9ca8e5.png)

chartWidth 50 : 1,25 Ko (taille 3,07 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288490-45a7b280-37f7-11eb-8521-92b0b6facd8f.png)

- LTD:
chartWidth 1000 : 4,71 Ko (taille 39,74 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288542-7c7dc880-37f7-11eb-94e3-4fa7338f8f43.png)

chartWidth 500 : 2,93 Ko (taille 20,42 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288546-7ee02280-37f7-11eb-9785-43459df1d02f.png)

chartWidth 250 : 1,85 Ko (taille 10,79 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288552-81db1300-37f7-11eb-9c7d-f774c825a37c.png)

chartWidth 100 : 1,43 Ko (taille 5,00 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288555-843d6d00-37f7-11eb-858d-1c65d20cc1f7.png)

chartWidth 50 : 1,24 Ko (taille 3,06 Ko)
![image](https://user-images.githubusercontent.com/9742965/101288560-87d0f400-37f7-11eb-9653-60a07ec6f3b9.png)

- SMA:
chartWidth 1000 : 111,79 Ko (taille 1,21 Mo) => 43641 values returns
![image](https://user-images.githubusercontent.com/9742965/101288610-b5b63880-37f7-11eb-8e48-fb9a693ea289.png)

chartWidth 500 : 113,05 Ko (taille 1,22 Mo) => 44141 values returns
![image](https://user-images.githubusercontent.com/9742965/101288608-b353de80-37f7-11eb-8db8-1906e25d15a1.png)

chartWidth 250 : 113,76 Ko (taille 1,23 Mo) => 44391 values returns
![image](https://user-images.githubusercontent.com/9742965/101288601-b0f18480-37f7-11eb-8ee5-10c5c292e53e.png)

chartWidth 100 : 114,16 Ko (taille 1,23 Mo) => 44541 values returns
![image](https://user-images.githubusercontent.com/9742965/101288595-ae8f2a80-37f7-11eb-9d17-2513cf42bda8.png)

chartWidth 50 : 114,31 Ko (taille 1,23 Mo) => 44591 values returns
![image](https://user-images.githubusercontent.com/9742965/101288592-ac2cd080-37f7-11eb-84c4-9cb2ef23e40a.png)

- ASAP: 
chartWidth 1000 : 4,97 Ko (taille 40,78 Ko) => 1014 values returns
![image](https://user-images.githubusercontent.com/9742965/101288636-d4b4ca80-37f7-11eb-82b9-9ae8945b9855.png)

chartWidth 500 : 3,39 Ko (taille 20,45 Ko) => 470 values returns
![image](https://user-images.githubusercontent.com/9742965/101288633-d2527080-37f7-11eb-9322-4245be2bdae8.png)

chartWidth 250 : 2,24 Ko (taille 10,67 Ko) => 233 values returns
![image](https://user-images.githubusercontent.com/9742965/101288630-cf578000-37f7-11eb-96ed-7b8b13b396a0.png)

chartWidth 100 : 1,65 Ko (taille 5,26 Ko) => 100 values returns
![image](https://user-images.githubusercontent.com/9742965/101288628-ccf52600-37f7-11eb-98de-ffdb33e5a72d.png)

chartWidth 50 : 1,41 Ko (taille 3,03 Ko) => 46 values returns
![image](https://user-images.githubusercontent.com/9742965/101288625-ca92cc00-37f7-11eb-88f3-cf10c801cb4c.png)

### Conclusions:
Performance (weight of result only):
=> (approximately) 1000 values = 6 ko ;  500 values = 3k ; 250 = 2.5ko ; 100 = 1.5 ; 50 = 1.3
=> I will try to limit  the size (limit value to 1 decimal, check date format, remove device or features informations not really need, any others idea ?)
=> In my opinion, 100 values returns is the good numbers

Visualization:
=> In my opinion, with those tests, I think LTD is the best to see a good representation of dataset, you see min max value and variation is not each time invisible 
=> LTOB and ASAP are to far of real dataset
=> I want play more with SMA to check others parameter 
=> I want compare with more "light" algo, for example extract only min max by day (maybe I can reduce weight with same visualization) 
=> Another solution can be to display different drill on X axis (=date), for example display min max by week if the selector is on month or year , it is more complicated for dev in front side and I'm not sure correspond to each use case

**Comment by Hamtar0** [comment]
And on mouse over, we can see values and dates ?

**Comment by Pierre-Gilles** [comment]
@euguuu Thanks for this very complete analysis! 

**Size:** I'm ok with 100 values. If the number of values recorded in the interval is less, you agree that the downsampling won't be needed right? 
**Performance:** How fast is the backend to go through the whole controller ? (GET SQL + downsampling?)

> I want compare with more "light" algo, for example extract only min max by day (maybe I can reduce weight with same visualization)

I think you'll have better results with those algorithm,+ you'll have to think of every filtering possible (by hour, by 24h, by week, by month) don't waste too much time on this, LTD is great :)

**Comment by euguuu** [comment]
> And on mouse over, we can see values and dates ?

On tooltip you see date and value but only on 100 values go up

**Comment by euguuu** [comment]
> @euguuu Thanks for this very complete analysis!
> 
> **Size:** I'm ok with 100 values. If the number of values recorded in the interval is less, you agree that the downsampling won't be needed right?
=> yes is alerady the case

> **Performance:** How fast is the backend to go through the whole controller ? (GET SQL + downsampling?)
> 
=> I check it latter

> I think you'll have better results with those algorithm,+ you'll have to think of every filtering possible (by hour, by 24h, by week, by month) don't waste too much time on this, LTD is great :)
=> ok

**Comment by euguuu** [comment]
**Performance:** How fast is the backend to go through the whole controller ? (GET SQL + downsampling?)
=> the time on server side is ~ 5s for 100 values (downsampling LTD ) 
=> I reduce weight of 100 values to 1Ko

**Comment by VonOx** [comment]
@euguuu seconds or milliseconds?

**Comment by euguuu** [comment]
is 5 seconds, it's long, how can I build a raspberry img to test on raspberry ?

I remove some padding in box:
![image](https://user-images.githubusercontent.com/9742965/102016295-712f1d80-3d60-11eb-973e-41e2b777899b.png)

about trend, I'm not sure comparison between 2 values describe a real trend, I think we need to take all data values display .
hat do you think about those lib:
https://www.npmjs.com/package/basic-trend
https://www.npmjs.com/package/regression
https://www.npmjs.com/package/simple-statistics
https://www.npmjs.com/package/regression-trend

**Comment by Pierre-Gilles** [comment]
> Performance: How fast is the backend to go through the whole controller ? (GET SQL + downsampling?)
=> the time on server side is ~ 5s for 100 values (downsampling LTD )

5 seconds? wow, it's way too long. It expected something like 100ms maximum. What's slow? The SQL query? The downsampling? 

Let's investigate what's wrong because 5 sec is way too long.

> about trend, I'm not sure comparison between 2 values describe a real trend, I think we need to take all data values display .
hat do you think about those lib:

Why? This is how most trading app works. See for example: https://finance.zacks.com/calculate-daily-price-variation-stocks-8299.html

From that link: 

> if a stock opened at $75 per share, fell to $70 and remained there until the end of the day, you would divide $5 (the daily price variation) by $70 (the closing, or current, price) for a result of .071, representing a 7.1% daily price variation.

Let's not use a library to display something as simple as a variation :)

**Comment by euguuu** [comment]
I work on performance aspect, 
I test on database with 100 800 rows repartee on ~ 2 month
1 month of data represent 44 640 rows

Before changes: 
- Total Time : 6875 ms 
- SQL time : 3170 ms 
- Downsampling time : 3705 ms

After changes:
- Total Time : 769 ms 
- SQL time : 744 ms => remove some attribute in select, do 2 different request (1 for get device/room/feature info and 1 for feature state) instead of 1 request with join 
- Downsampling time : 25 ms => change algo to LTTB 

(I change trend method to use last/first value without performance impact)

I try to find a solution to increase performance of sql request used to extract feature state ...

how can I build a raspberry img from my branch to test on raspberry ?

**Comment by Pierre-Gilles** [comment]
Very nice improvements, congrats 👏

I'm sure we can go even further in the SQL request :) What's the current SQL request?

> how can I build a raspberry img from my branch to test on raspberry ?

You can use the GitHub Action in the repo (https://github.com/GladysAssistant/Gladys/blob/master/.github/workflows/docker-dev-build.yml) 

You need to set   DOCKERHUB_USER,  DOCKERHUB_PASSWORD and change DOCKERHUB_REPO, then you can trigger the workflow and it'll build and push the image to your DockerHub account with the `:dev` tag. 

Edit: Or you could do it manually using the same command as in the GitHub action script, but locally on your computer :)

**Comment by euguuu** [comment]
there are 2 requests:
first (~ 10ms): 
SELECT
	`t_device`.`id`,
	`t_device`.`name`,
	`t_device`.`selector`,
	`features`.`id` AS `features.id`,
	`features`.`name` AS `features.name`,
	`features`.`selector` AS `features.selector`,
	`features`.`unit` AS `features.unit`,
	`features`.`last_value` AS `features.last_value`,
	`features`.`last_value_changed` AS `features.last_value_changed`,
	`room`.`id` AS `room.id`,
	`room`.`name` AS `room.name`,
	`room`.`selector` AS `room.selector`
FROM
	`t_device` AS `t_device`
INNER JOIN `t_device_feature` AS `features` ON
	`t_device`.`id` = `features`.`device_id`
	AND `features`.`selector` = 'test-temperature-sensor-2'
LEFT OUTER JOIN `t_room` AS `room` ON
	`t_device`.`room_id` = `room`.`id`
LIMIT 0,
10000000000000

seconds (~ 700 ms): 
SELECT
	`device_feature_id`,
	`value`,
	`created_at`
FROM
	`t_device_feature_state` AS `t_device_feature_state`
WHERE
	`t_device_feature_state`.`device_feature_id` = 'f07c5b27-9301-4482-a059-9f91329d30e7'
	AND `t_device_feature_state`.`created_at` BETWEEN '2020-12-09 15:19:26.160 +00:00' AND '2021-01-09 15:19:26.160 +00:00'
ORDER BY
	`t_device_feature_state`.`created_at` ASC

I put those 2 request in same transaction and 
I change index I put in t_device_feature_state to put a new index on column device_feature_id + created_at to improve explain plan
Now explain plan is: 
SEARCH TABLE t_device_feature_state AS t_device_feature_state USING INDEX t_device_feature_state_device_feature_id_created_at (device_feature_id=? AND created_at>? AND created_at<?) 
=> only one line in explain, use index for where and order by
=> those change win a couple ms but not significant

So I change sequelize call of request 2 (because I think it's more sequelize pb, when I execute this request in my sql client I have ~2 ms of exec and ~ 180ms to display)
now I use ~ same query but in raw query 
=> I think is best solution (is often the case with ORM) but not in code design, do you agree with raw query usage in gladys (i do not find any occurrence of this method in another request)? 
=> This method improve performance and now I have result in ~220ms ( total time : ~ 300ms)

Now I don't know how I can improve performance (I try soon on raspberry to know time in real target).

I find those articles : 
https://stackoverflow.com/questions/41965054/best-data-type-for-sqlite-index
https://stackoverflow.com/questions/17277735/using-uuids-in-sqlite
https://mareks-082.medium.com/auto-increment-keys-vs-uuid-a74d81f7476a
=> maybe UUID column do not help in index performance but I'm not sure about that and is big change
=> I do not find any information about different type of index, do you know if SQLite have another index type ? (in some database btree index on datetime column is not the best)
=> I think the time is lose in serialize data and not in sql request execution, do you know if it's possible to fix fetch size with sequelize ? (I'm not sure this config exist in javascript database driver but this config often increase fetch in Java)

**Comment by euguuu** [comment]
I have this error when I try to build img, what I do wrong ?:
![image](https://user-images.githubusercontent.com/9742965/104129377-65894380-536c-11eb-9799-5ec9e4f2de97.png)

**Comment by VonOx** [comment]
Because DOCKERHUB_USER AND/OR DOCKERHUB_PASSWORD is not set in repo secrets.

**Comment by Pierre-Gilles** [comment]
@euguuu Nice investigation! Are you sure that sequelize produce a different SQL code than your raw SQL ? If possible, I would like to avoid writing raw SQL, because it'll make it difficult in a near future to configure Gladys with different DB engine (postgresql, mysql, ...) 

Your SQL query is really simple and I'm sure we could have the same query with Sequelize! What's your sequelize code?

**Comment by euguuu** [comment]
sorry for misunderstood, 
sql reqest is same with sequelize (i just remove some quote en table prefix)

but when you do this request with sequelize the json object contain more information (about transaction, and others things but not in real content) and with raw we have directly the content (i do not need to do second action plain to get object) 

I send you code after, but clearly sql request is the same (and I think perfomance is not loose in sql execution)

PS: I'm very happy to see we can think to change with most powerfull database like postgres (in this case datbase engine change do not impact raw query)

**Comment by Pierre-Gilles** [comment]
Hi! Just found this today, have you tried this: 

<img width="1164" alt="Screenshot 2021-01-29 at 17 01 28" src="https://user-images.githubusercontent.com/7365207/106298109-c57c5700-6253-11eb-8c96-674ba25bf8ec.png">

**Comment by euguuu** [comment]
I apply a downsampling of state in periodic job

now for display last month of date (1 state by minute):
=> return 3011 raw from state_light and 1300 from state
=> total time of call ~ 80ms (70 ms sql + 10 downsampling)

**Comment by Pierre-Gilles** [comment]
Ok, and where do you store them at the end?

**Comment by euguuu** [comment]
I store state after downsamling in new table t_device_feature_state_light

**Comment by Pierre-Gilles** [comment]
Ok! And how often do you build this "cached" version ?

**Comment by euguuu** [comment]
every day

**Comment by Pierre-Gilles** [comment]
what happens when it's the first day?

**Comment by euguuu** [comment]
i add a column last_downsampling on feature object

the first time of job execution the donwsampling start on all data in feature_state and the second time just on last day of last_downsampling (=last created_at of feature_state treat) in feature,
so if some feature state arrive in late (for example if data come from third party platform and this platform is down for a couple of days) the downsampling treat all data from last_downsampling date (so maybe more than 1 day)

**Comment by Pierre-Gilles** [comment]
FYI: https://community.gladysassistant.com/t/affichage-des-courbes-de-capteurs-sur-le-dashboard/6535
